### PR TITLE
Fix None torrent_id assertion

### LINKED ISSUE #1426: exceptions.AssertionError:  in addOrGetTorrentIDSReturn
Latest devel, with tunnels enabled.
```
ERROR   1432082812.87 RemoteTorrentHandler:600   failed to save data for download fbbd64b31eb3de5b546995e6f61aeb892968bc8f.torrent: info misses key pieces
ERROR   1432086838.89             log:455   Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/pouwelse/GITHUB/tribler/Tribler/Core/Utilities/twisted_thread.py", line 54, in _reactor_runner
    reactor.run(installSignalHandlers=False)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1192, in run
    self.mainLoop()
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 1201, in mainLoop
    self.runUntilCurrent()
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 824, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "/home/pouwelse/GITHUB/tribler/Tribler/dispersy/community.py", line 2050, in _process_message_batch
    return self._on_batch_cache(meta, batch)
  File "/home/pouwelse/GITHUB/tribler/Tribler/dispersy/community.py", line 2092, in _on_batch_cache
    self.on_messages(messages)
  File "/home/pouwelse/GITHUB/tribler/Tribler/dispersy/community.py", line 2206, in on_messages
    if self._dispersy.store_update_forward(possibly_messages, True, True, False):
  File "/home/pouwelse/GITHUB/tribler/Tribler/dispersy/dispersy.py", line 1694, in store_update_forward
    if self._update(possibly_messages) == False:
  File "/home/pouwelse/GITHUB/tribler/Tribler/dispersy/util.py", line 151, in wrapper
    return_value = func(*args, **kargs)
  File "/home/pouwelse/GITHUB/tribler/Tribler/dispersy/dispersy.py", line 1719, in _update
    messages[0].handle_callback(messages)
  File "/home/pouwelse/GITHUB/tribler/Tribler/community/channel/community.py", line 391, in _disp_on_torrent
    self._channelcast_db.on_torrents_from_dispersy(torrentlist)
  File "/home/pouwelse/GITHUB/tribler/Tribler/Core/CacheDB/SqliteCacheDBHandler.py", line 1552, in on_torrents_from_dispersy
    torrent_ids, inserted = self.torrent_db.addOrGetTorrentIDSReturn(infohashes)
  File "/home/pouwelse/GITHUB/tribler/Tribler/Core/CacheDB/SqliteCacheDBHandler.py", line 503, in addOrGetTorrentIDSReturn
    assert all(torrent_id for torrent_id in torrent_ids), torrent_ids
exceptions.AssertionError: [308452, None, 25260, 308455, 308454, 308457, 308451, 308456, 308450, 308453, 308458, 308449, 308459]
```

### PR DESCRIPTION
**Pull request by LipuFei**
Closes #1426

### CODE REVIEW

**Comment by synctext** [comment]
retest this please

File: Tribler/Core/CacheDB/SqliteCacheDBHandler.py [hunk 1]
```diff
@@ -492,6 +492,9 @@ def addOrGetTorrentIDSReturn(self, infohashes):
         for infohash, torrent_id in torrent_id_results.iteritems():
             if torrent_id is None:
                 to_be_inserted.add(infohash)
+        for infohash in infohashes:
```
Comment (NielsZeilemaker) [inline Tribler/Core/CacheDB/SqliteCacheDBHandler.py]: This is a hack, getTorrentIDS should return a list consisting of infohash, torrent_id for all infohashes in infohashes.
Fix getTorrentIDS instead.

**Comment by LipuFei** [comment]
@NielsZeilemaker I think I found out why. The `torrent_id` cache doesn't have the infohash when this issue happens. The `torrent_id` cache is a `LimitedOrderedDict`. In `getTorrentIDS()`, we first check the cache to see which infohashes do we need to select from the database, then we select, feed the cache, and get the torrent_ids from the cache. I think when we were feeding the cache, it got full and started overwriting the existing infohashes, and it accidentally overwrote one or more of the infohashes we need to get. So when it tries to get torrent_ids from the cache in the end, it couldn't get everything.

**Comment by NielsZeilemaker** [comment]
Nice
Op 22 mei 2015 11:40 AM schreef "Lipu Fei" <notifications@github.com>:

**Comment by LipuFei** [comment]
@whirm @NielsZeilemaker ready

**Comment by LipuFei** [comment]
@NielsZeilemaker @whirm Ready after this one

**Comment by LipuFei** [comment]
ready

**Comment by whirm** [comment]
looks fine to me

**Pull request merged**